Add cross-dev and cross-util commands.

Alexhuszagh opened the pull request:
The cross-dev command has the `target-info` subcommand, which target information (libc, compiler, C++, qemu version) for a given target.

```bash
cross-dev target-info
cross-dev target-info arm-unknown-linux-gnueabihf
```

The cross-util command has the `list-images` and `remove-images` subcommands. `list-images` lists all cross images, including local development files. `remove-images` removes cross images.

```bash
cross-util list-images
cross-util remove-images
cross-util remove-images arm-unknown-linux-gnueabihf
```

Closes #760.

Alexhuszagh commented:
I've separated this into many different commits, since this is a large PR and refactors some logic. Each I think is a relatively independent commit, and some of them (such as the removal of the duplicate `cpp` tags) is crucial for subsequent steps (`serde_yaml` parsing the matrix).

Emilgardis commented on src/bin/cross-util.rs:
```diff
@@ -0,0 +1,193 @@
+        local: bool,
+        /// Container engine (such as docker or podman).
+        #[clap(long)]
+        engine: Option<String>,
```
this and the previous one should also be a env var `CROSS_CONTAINER_ENGINE`

Emilgardis replied:
https://docs.rs/clap/3.1.18/clap/struct.Arg.html#method.env

Alexhuszagh replied:
How would this work then with `cross::get_container_engine`?

Emilgardis replied:
ah, I see, nevermind then :D

Emilgardis commented on src/bin/cross-util.rs:
```diff
@@ -0,0 +1,193 @@
+}
+
+fn is_local_image(image: &str) -> bool {
+    // this cannot panic: we've formatted our image as ${repo}:${tag}
```
```suggestion
    // this cannot panic: we've formatted our image list as ${repo}:${tag}
```

Emilgardis commented on src/bin/cross-util.rs:
```diff
@@ -0,0 +1,193 @@
+            engine,
+        } => {
+            let engine = get_container_engine(engine)?;
+            if targets.is_empty() {
+                remove_all_images(&engine, *verbose, *force, *local)?;
+            } else {
+                remove_target_images(&engine, targets, *verbose, *force, *local)?;
+            }
```
i think there should be a confirm prompt here, or, a we have a `--execute` flag and standard operation without it is a dry-run

Emilgardis commented on src/bin/cross-util.rs:
```diff
@@ -0,0 +1,193 @@
+    },
+}
+
+fn get_container_engine(engine: &Option<String>) -> Result<PathBuf, which::Error> {
```
```suggestion
fn get_container_engine(engine: Option<&str>) -> Result<PathBuf, which::Error> {
```

`Option` has a `as_ref` implementation

Emilgardis commented on src/bin/cross-util.rs:
```diff
@@ -0,0 +1,193 @@
+//  repository = (${registry}/)?rustembedded/cross
+//  tag = ${target}(-${version})?
+// our target triple must match `[A-Za-z0-9_-]`
+fn rustembedded_target(tag: &str) -> Option<String> {
```
why does this return an option?

Alexhuszagh replied:
It should not. I had originally planned some checking if the triple had <3 components, but I don't think that's a great idea anymore (say, a custom image someone also wants to remove).

Emilgardis commented on Cargo.toml:
```diff
@@ -8,9 +8,11 @@ name = "cross"
 repository = "https://github.com/cross-rs/cross"
 version = "0.2.1"
 edition = "2021"
+readme = "docs/crates_io.md"
```
why?

Alexhuszagh replied:
This is so when published on `crates.io`, it ensures that the README will display [this](https://github.com/Alexhuszagh/cross/blob/14857a333d6ad81e04b0af89aa9a3573bca07193/docs/crates_io.md). On `docs.rs`, it will look similar to the following image:

![Screenshot_20220608_091729](https://user-images.githubusercontent.com/9440903/172639769-8d3e2913-a33b-419a-ab89-a1332455352f.png)

Emilgardis replied:
the warning would be nice to have in the documentation, but for crates.io it's fine to only have the regular readme I think.

It's not needed to specify README.md as the readme in the manifest, see https://github.com/cross-rs/cross/pull/563

Emilgardis commented on cross-dev/Cargo.lock:
```diff
@@ -0,0 +1,621 @@
+# This file is automatically @generated by Cargo.
```
we should not have multiple lock files imo, instead `cross-dev` should be included in the root workspace, preferably also then without being in the default-members

Emilgardis replied:
no, sadly you can't. :/

you can't `use bin;` if `bin` is a binary target, you need to have a `lib.rs` to use it in other crates, even if those crates are in the same workspace

Emilgardis commented on cross-dev/Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+name = "cross-dev"
+repository = "https://github.com/cross-rs/cross"
+version = "0.0.1"
+edition = "2021"
+readme = "../docs/crates_io.md"
```
```suggestion
edition = "2021"
publish = false
```

Emilgardis commented on cross-dev/Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+
+[dependencies]
+clap = { version = "3.1.18", features = ["derive"] }
+cross = { path = "..", version = "0.2.1" }
```
```suggestion
cross = { path = ".." }
```

Emilgardis commented on cross-dev/src/main.rs:
```diff
@@ -0,0 +1,208 @@
+#[derive(Debug, PartialEq, Deserialize)]
+struct Matrix {
+    target: String,
+    os: String,
+    #[serde(default)]
+    cpp: i64,
+    #[serde(default)]
+    std: i64,
+    #[serde(default)]
+    run: i64,
+    runners: Option<String>,
```
are these values used?

Alexhuszagh replied:
They are not: I considered if they would and they should be removed.

Alexhuszagh commented:
Is this better now? I incorporated some of the changes, but thought it would be a bit better to provide a `[[bin]]` target for `cross-dev`, since then it can be run easily but disabled by default.

The changes I've made include:
- Moving `cross-dev` to `/cross-dev/main.rs`, and using the main `Cargo.toml` for it. No more `Cargo.lock`, no more need to not publish it.
- Make `cross-dev` a `[[bin]]` target.Since we need `serde_yaml` as a dependency for it, I've changed it so it's only compiled if the `dev` feature is enabled.
- `engine` still uses the same logic. I considered using `default_value_t` and `parse` for the clap options, but it was going to be a lot of code and would have probably been less readable than the existing implementation. This is because we'd have to provide a wrapper for `which::which`, then use `default_value_t = cross::get_container_engine()`.
- Changed the image output to `{{.Repository}}:{{.Tag}} {{.ID}}`. This is then parsed to an `Image` (see below). This is because images may be tagged as `${repository}:<none>`, which would cause on error on removing the image. Removing it by ID works in this case.
- I've added the `--execute` flag, so it's a dry-run by default.
- I've removed the unnecessary `Matrix` fields.
- I've changed `get_container_engine` to take an `Option<&str>`.
- `rustembedded_target` no longer returns an option.
- No longer use a custom readme in `Cargo.toml`. `lib.rs` still warns about this for crates.io.

```rust
#[derive(Debug, PartialOrd, Ord, PartialEq, Eq)]
struct Image {
    repository: String,
    tag: String,
    // need to remove images by ID, not just tag
    id: String,
}
```

Emilgardis commented on docs/crates_io.md:
```diff
@@ -0,0 +1,11 @@
+[![crates.io](https://img.shields.io/crates/v/cross.svg)](https://crates.io/crates/cross)
```
this file should be renamed, it can even be ommited fully and just put in the lib.rs directly

Emilgardis commented on src/lib.rs:
```diff
@@ -1,3 +1,4 @@
+#![doc = include_str!("../docs/crates_io.md")]
```
as mentioned, this could be inlined

Emilgardis commented on src/bin/cross-util.rs:
```diff
@@ -0,0 +1,229 @@
+    } else if let Some(tag) = image.tag.strip_prefix(DOCKER_IO) {
+        Ok(rustembedded_target(tag))
+    } else {
+        eyre::bail!("Cannot get target for image {}", image.name())
```
```suggestion
        eyre::bail!("cannot get target for image {}", image.name())
```

Emilgardis commented on .github/workflows/ci.yml:
```diff
@@ -249,6 +249,10 @@ jobs:
 
+      - name: Install cross-dev
+        if: matrix.deploy
+        run: cargo install --path cross-dev --force
```
specifying `--bin cross-dev` should be done too, to avoid unsync between needed features and the ci stage

Emilgardis reviewed (approved):
bors r+

Alexhuszagh commented:
> bors r-
> 
> oh yeah you wanted to squash?

Already did.

bors r=Emilgardis
